File: shared/RemodexSharedCore/Sources/RemodexSharedCore/Models/RPCMessage.swift
import Foundation

public typealias RPCObject = [String: JSONValue]

public struct RPCMessage: Codable, Sendable {
    public let jsonrpc: String?
    public let id: JSONValue?
    public let method: String?
    public let params: JSONValue?
    public let result: JSONValue?
    public let error: RPCError?

    public init(
        id: JSONValue? = nil,
        method: String,
        params: JSONValue? = nil,
        includeJSONRPC: Bool = true
    ) {
        self.jsonrpc = includeJSONRPC ? "2.0" : nil
        self.id = id
        self.method = method
        self.params = params
        self.result = nil
        self.error = nil
    }

    public init(
        id: JSONValue?,
        result: JSONValue,
        includeJSONRPC: Bool = true
    ) {
        self.jsonrpc = includeJSONRPC ? "2.0" : nil
        self.id = id
        self.method = nil
        self.params = nil
        self.result = result
        self.error = nil
    }

    public init(
        id: JSONValue?,
        error: RPCError,
        includeJSONRPC: Bool = true
    ) {
        self.jsonrpc = includeJSONRPC ? "2.0" : nil
        self.id = id
        self.method = nil
        self.params = nil
        self.result = nil
        self.error = error
    }

    public init(
        jsonrpc: String? = "2.0",
        id: JSONValue? = nil,
        method: String? = nil,
        params: JSONValue? = nil,
        result: JSONValue? = nil,
        error: RPCError? = nil
    ) {
        self.jsonrpc = jsonrpc
        self.id = id
        self.method = method
        self.params = params
        self.result = result
        self.error = error
    }
}

public extension RPCMessage {
    var isRequest: Bool {
        method != nil
    }

    var isResponse: Bool {
        result != nil || error != nil
    }

    var isErrorResponse: Bool {
        error != nil
    }
}

public struct RPCError: Codable, Error, Sendable {
    public let code: Int
    public let message: String
    public let data: JSONValue?

    public init(code: Int, message: String, data: JSONValue? = nil) {
        self.code = code
        self.message = message
        self.data = data
    }
}

File: android/app/src/main/kotlin/com/remodex/mobile/core/model/RpcMessage.kt
package com.remodex.mobile.core.model

import kotlinx.serialization.Serializable

@Serializable
data class RPCMessage(
    val jsonrpc: String? = null,
    @Serializable(JSONValueSerializer::class)
    val id: JSONValue? = null,
    val method: String? = null,
    @Serializable(JSONValueSerializer::class)
    val params: JSONValue? = null,
    @Serializable(JSONValueSerializer::class)
    val result: JSONValue? = null,
    val error: RPCError? = null,
) {
    val isRequest: Boolean get() = method != null

    val isResponse: Boolean get() = result != null || error != null

    val isErrorResponse: Boolean get() = error != null

    companion object {
        fun request(
            id: JSONValue? = null,
            method: String,
            params: JSONValue? = null,
            includeJsonRpc: Boolean = true,
        ): RPCMessage =
            RPCMessage(
                jsonrpc = if (includeJsonRpc) "2.0" else null,
                id = id,
                method = method,
                params = params,
                result = null,
                error = null,
            )

        fun success(
            id: JSONValue?,
            result: JSONValue,
            includeJsonRpc: Boolean = true,
        ): RPCMessage =
            RPCMessage(
                jsonrpc = if (includeJsonRpc) "2.0" else null,
                id = id,
                method = null,
                params = null,
                result = result,
                error = null,
            )

        fun failure(
            id: JSONValue?,
            error: RPCError,
            includeJsonRpc: Boolean = true,
        ): RPCMessage =
            RPCMessage(
                jsonrpc = if (includeJsonRpc) "2.0" else null,
                id = id,
                method = null,
                params = null,
                result = null,
                error = error,
            )

        fun notification(
            method: String,
            params: JSONValue? = null,
            includeJsonRpc: Boolean = false,
        ): RPCMessage =
            RPCMessage(
                jsonrpc = if (includeJsonRpc) "2.0" else null,
                id = null,
                method = method,
                params = params,
                result = null,
                error = null,
            )
    }
}

@Serializable
data class RPCError(
    val code: Int,
    val message: String,
    @Serializable(JSONValueSerializer::class)
    val data: JSONValue? = null,
)

/**
 * True when the bridge reports the thread is gone (parity iOS `shouldTreatAsThreadNotFound`),
 * excluding "not materialized yet" transient errors that share [-32600].
 */
fun RPCError.isExplicitServerThreadMissing(): Boolean {
    val m = message.lowercase()
    if (m.contains("not materialized") || m.contains("not yet materialized")) return false
    return m.contains("thread not found") || m.contains("unknown thread")
}
